### PR TITLE
tabletserver: fix flaky test

### LINKED ISSUE #7832: Flaky TestConsolidatorMemoryLimits
#### Overview of the Issue
--- FAIL: TestConsolidatorMemoryLimits (0.54s)
    --- FAIL: TestConsolidatorMemoryLimits/multiple_phase_consolidation (0.16s)
        stream_consolidator_test.go:387: 
            	Error Trace:	stream_consolidator_test.go:387
            	Error:      	Not equal: 
            	            	expected: 4
            	            	actual  : 3
            	Test:       	TestConsolidatorMemoryLimits/multiple_phase_consolidation

### LINKED ISSUE #7837: Test Failing TestConsolidatorMemoryLimits
```
021-04-12T19:19:34.1415616Z ok  	vitess.io/vitess/go/vt/vttablet/tabletmanager	0.280s
2021-04-12T19:19:34.6569753Z W0412 19:19:26.807820   26560 parser.go:83] ignoring error parsing DDL 'create index a on user(id1 + id2)': syntax error at position 29
2021-04-12T19:19:34.6572436Z W0412 19:19:26.808954   26560 parser.go:83] ignoring error parsing DDL 'create index a on user(id1 + id2)': syntax error at position 29
2021-04-12T19:19:34.6573947Z W0412 19:19:26.841273   26560 tabletserver.go:1361] Code: ABORTED
2021-04-12T19:19:34.6575507Z transaction 1618255166839506972: ended at 2021-04-12 19:19:26.841 UTC (closed) (CallerID: d)
2021-04-12T19:19:34.6576301Z : Sql: "commit", BindVars: {}
2021-04-12T19:19:34.6577014Z W0412 19:19:26.844273   26560 tabletserver.go:1361] Code: ABORTED
2021-04-12T19:19:34.6578124Z transaction 1618255166842399696: ended at 2021-04-12 19:19:26.844 UTC (closed) (CallerID: d)
2021-04-12T19:19:34.6578991Z : Sql: "commit", BindVars: {}
2021-04-12T19:19:34.6579916Z E0412 19:19:26.860894   26560 message_manager.go:384] Terminating connection due to error sending field info: EOF
2021-04-12T19:19:34.6581184Z E0412 19:19:26.894379   26560 querylogz.go:126] unexpected value in TxLog: "test msg" (expecting value of type *tabletenv.LogStats)
2021-04-12T19:19:34.6582570Z E0412 19:19:26.920271   26560 state_manager.go:276] Error transitioning to the desired state: MASTER, Serving, will keep retrying: intentional error
2021-04-12T19:19:34.6583904Z E0412 19:19:27.172564   26560 state_manager.go:276] Cannot connect to MySQL, shutting down query service: intentional error
2021-04-12T19:19:34.6585289Z --- FAIL: TestConsolidatorMemoryLimits (0.54s)
2021-04-12T19:19:34.6586741Z     --- FAIL: TestConsolidatorMemoryLimits/two-phase_consolidation_(memory) (0.02s)
2021-04-12T19:19:34.6588001Z         stream_consolidator_test.go:354: 
2021-04-12T19:19:34.6588811Z             	Error Trace:	stream_consolidator_test.go:354
2021-04-12T19:19:34.6589544Z             	Error:      	Not equal: 
2021-04-12T19:19:34.6590161Z             	            	expected: 2
2021-04-12T19:19:34.6590906Z             	            	actual  : 1
2021-04-12T19:19:34.6592249Z             	Test:       	TestConsolidatorMemoryLimits/two-phase_consolidation_(memory)
2021-04-12T19:19:34.6593566Z E0412 19:19:29.912379   26560 tx_engine.go:363] Error extracting transaction ID from ditd: Code: INVALID_ARGUMENT
2021-04-12T19:19:34.6594466Z invalid parts in dtid: dtid0
2021-04-12T19:19:34.6595355Z E0412 19:19:29.913626   26560 tx_engine.go:363] Error extracting transaction ID from ditd: Code: INVALID_ARGUMENT
2021-04-12T19:19:34.6596239Z invalid parts in dtid: bogus
2021-04-12T19:19:34.6597714Z E0412 19:19:29.913977   26560 tx_engine.go:177] Could not prepare transactions: unknown error: query: 'bogus' is not supported on fakesqldb (errno 1105) (sqlstate HY000) during query: bogus
2021-04-12T19:19:34.6598945Z E0412 19:19:29.921348   26560 tabletserver.go:1361] Code: NOT_FOUND
2021-04-12T19:19:34.6599715Z could not transition to COMMIT: aa
2021-04-12T19:19:34.6600382Z : Sql: "start_commit", BindVars: {}
2021-04-12T19:19:34.6602161Z E0412 19:19:29.924430   26560 tabletserver.go:1361] Code: NOT_FOUND
2021-04-12T19:19:34.6602980Z could not transition to ROLLBACK: aa
2021-04-12T19:19:34.6603639Z : Sql: "set_rollback", BindVars: {}
2021-04-12T19:19:34.6604478Z E0412 19:19:29.934473   26560 tabletserver.go:1361] PoolFull: Code: RESOURCE_EXHAUSTED
2021-04-12T19:19:34.6605541Z transaction pool aborting request due to already expired context
2021-04-12T19:19:34.6606309Z : Sql: "begin", BindVars: {}
2021-04-12T19:19:34.6607020Z W0412 19:19:29.939099   26560 tabletserver.go:1361] Code: ABORTED
2021-04-12T19:19:34.6608350Z transaction -1: not found
2021-04-12T19:19:34.6609006Z : Sql: "commit", BindVars: {}
2021-04-12T19:19:34.6609708Z W0412 19:19:29.939136   26560 tabletserver.go:1361] Code: ABORTED
2021-04-12T19:19:34.6610636Z transaction -1: not found
2021-04-12T19:19:34.6611271Z : Sql: "rollback", BindVars: {}
2021-04-12T19:19:34.6612002Z W0412 19:19:29.960112   26560 tabletserver.go:1361] Code: ABORTED
2021-04-12T19:19:34.6612725Z transaction 123456: not found
2021-04-12T19:19:34.6613347Z : Sql: "select 42", BindVars: {}
2021-04-12T19:19:34.6614198Z W0412 19:19:29.962461   26560 tabletserver.go:1361] Code: ABORTED
2021-04-12T19:19:34.6614906Z transaction 123456: not found
2021-04-12T19:19:34.6615469Z : Sql: "", BindVars: {}
2021-04-12T19:19:34.6616524Z E0412 19:19:29.966974   26560 tabletserver.go:1361] unknown error: it broke (errno 1105) (sqlstate HY000): Sql: "begin", BindVars: {}
2021-04-12T19:19:34.6617501Z W0412 19:19:29.969387   26560 tabletserver.go:1361] Code: ABORTED
2021-04-12T19:19:34.6627256Z transaction 1618255169967758478: ended at 2021-04-12 19:19:29.969 UTC (release connection)
2021-04-12T19:19:34.6628264Z : Sql: "", BindVars: {}
2021-04-12T19:19:34.6628956Z W0412 19:19:29.969420   26560 tabletserver.go:1361] Code: ABORTED
2021-04-12T19:19:34.6631043Z transaction 1618255169967758478: ended at 2021-04-12 19:19:29.969 UTC (release connection)
2021-04-12T19:19:34.6631987Z : Sql: "", BindVars: {}
2021-04-12T19:19:34.6633122Z E0412 19:19:29.992385   26560 tabletserver.go:1361] unknown error: rejected (errno 1105) (sqlstate HY000): Sql: "insert into test_table values (1, 2)", BindVars: {}
2021-04-12T19:19:34.6634680Z E0412 19:19:30.025803   26560 tabletserver.go:1361] Code: CANCELED
2021-04-12T19:19:34.6635375Z context canceled
2021-04-12T19:19:34.6636390Z : Sql: "waitForSameRangeTransactions", BindVars: {}
2021-04-12T19:19:34.6637607Z E0412 19:19:30.029550   26560 message_manager.go:384] Terminating connection due to error sending field info: EOF
2021-04-12T19:19:34.6642301Z E0412 19:19:30.033118   26560 tabletserver.go:1361] unknown error: query: 'update msg set time_acked = 1618255170032672506, time_next = null where id in ('1', '2') and time_acked is null limit 10001' is not supported on fakesqldb (errno 1105) (sqlstate HY000): Sql: "update msg set time_acked = :time_acked, time_next = null where id in ::ids and time_acked is null", BindVars: {#maxLimit: "type:INT64 value:\"10001\" "ids: "type:TUPLE values:<type:VARBINARY value:\"1\" > values:<type:VARBINARY value:\"2\" > "time_acked: "type:INT64 value:\"1618255170032672506\" "}
2021-04-12T19:19:34.6670185Z E0412 19:19:30.041214   26560 tabletserver.go:1361] unknown error: query: 'update msg set time_next = 1618255170040632671 + 30000000000 + if(FLOOR((30000000000 << ifnull(epoch, 0)) * 0.8226262344186889) < 30000000000, 30000000000, FLOOR((30000000000 << ifnull(epoch, 0)) * 0.8226262344186889)), epoch = ifnull(epoch, 0) + 1 where id in ('1', '2') and time_acked is null limit 10001' is not supported on fakesqldb (errno 1105) (sqlstate HY000): Sql: "update msg set time_next = :time_now + :wait_time + IF(FLOOR((:min_backoff<<ifnull(epoch, 0)) * :jitter) < :min_backoff, :min_backoff, FLOOR((:min_backoff<<ifnull(epoch, 0)) * :jitter)), epoch = ifnull(epoch, 0)+1 where id in ::ids and time_acked is null", BindVars: {#maxLimit: "type:INT64 value:\"10001\" "ids: "type:TUPLE values:<type:VARBINARY value:\"1\" > values:<type:VARBINARY value:\"2\" > "jitter: "type:FLOAT64 value:\"0.8226262344186889\" "min_backoff: "type:INT64 value:\"30000000000\" "time_now: "type:INT64 value:\"1618255170040632671\" "wait_time: "type:INT64 value:\"30000000000\" "}
2021-04-12T19:19:34.6675673Z E0412 19:19:30.045324   26560 tabletserver.go:1361] unknown error: query: 'delete from msg where time_acked < 0 limit 500' is not supported on fakesqldb (errno 1105) (sqlstate HY000): Sql: "delete from msg where time_acked < :time_acked limit 500", BindVars: {time_acked: "type:INT64 value:\"0\" "}
2021-04-12T19:19:34.6677503Z E0412 19:19:30.047517   26560 tabletserver.go:1277] Uncaught panic for Sql: "select * from test_table", BindVars: {}:
2021-04-12T19:19:34.6678218Z unknown exec error
2021-04-12T19:19:34.6679218Z /home/runner/work/vitess/vitess/go/vt/vttablet/tabletserver/tabletserver_test.go:1658 (0x10d02f2)
2021-04-12T19:19:34.6680860Z 	io/vitess/go/vt/vttablet/tabletserver.TestHandleExecUnknownError: panic("unknown exec error")
2021-04-12T19:19:34.6682320Z /opt/hostedtoolcache/go/1.15.11/x64/src/testing/testing.go:1123 (0x52c04e)
2021-04-12T19:19:34.6682957Z 	tRunner: fn(t)
2021-04-12T19:19:34.6683640Z /opt/hostedtoolcache/go/1.15.11/x64/src/runtime/asm_amd64.s:1374 (0x472440)
2021-04-12T19:19:34.6684595Z 	goexit: BYTE	$0x90	// NOP
2021-04-12T19:19:34.6685359Z E0412 19:19:30.048037   26560 tabletserver_test.go:1691] Code: INTERNAL
2021-04-12T19:19:34.6686091Z tablet error
2021-04-12T19:19:34.6686589Z : Sql: "select * from test_table", BindVars: {}
2021-04-12T19:19:34.6687221Z E0412 19:19:30.048486   26560 tabletserver_test.go:1691] Code: INTERNAL
2021-04-12T19:19:34.6687747Z tablet error
2021-04-12T19:19:34.6688239Z : Sql: "select * from test_table", BindVars: {}
2021-04-12T19:19:34.6699045Z E0412 19:19:30.049598   26560 tabletserver_test.go:1691] sensitive message (errno 10) (sqlstate HY000): Sql: "select * from test_table where a = :a", BindVars: {a: "type:INT64 value:\"1\" "}
2021-04-12T19:19:34.6700611Z E0412 19:19:30.050497   26560 tabletserver_test.go:1691] sensitive message (errno 10) (sqlstate HY000): Sql: "select * from test_table where xyz = :vt [TRUNCATED]", BindVars: {vtg1: "type:VARBINARY value:\ [TRUNCATED]
2021-04-12T19:19:34.6702482Z E0412 19:19:30.050528   26560 tabletserver_test.go:1691] sensitive message (errno 10) (sqlstate HY000): Sql: "select * from test_table where xyz = :vtg1 order by abc desc", BindVars: {vtg1: "type:VARBINARY value:\"this is kinda long eh\" "}
2021-04-12T19:19:34.6703899Z W0412 19:19:30.173148   26560 tabletserver.go:1361] Code: ABORTED
2021-04-12T19:19:34.6705211Z transaction 1618255170169830728: ended at 2021-04-12 19:19:30.173 UTC (release connection)
2021-04-12T19:19:34.6705792Z : Sql: "", BindVars: {}
2021-04-12T19:19:34.6706450Z E0412 19:19:30.215641   26560 twopc.go:289] Error parsing state for dtid dtid1: Code: INVALID_ARGUMENT
2021-04-12T19:19:34.6707257Z strconv.ParseInt: parsing "Failed": invalid syntax
2021-04-12T19:19:34.6707970Z .
2021-04-12T19:19:34.6708900Z E0412 19:19:30.215670   26560 twopc.go:289] Error parsing state for dtid dtid2: Code: INVALID_ARGUMENT
2021-04-12T19:19:34.6709910Z strconv.ParseInt: parsing "Failed": invalid syntax
2021-04-12T19:19:34.6710456Z .
2021-04-12T19:19:34.6711545Z W0412 19:19:30.321270   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255170219684936	'<nil>'	'<nil>'	Apr 12 19:19:30.219981	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6712897Z W0412 19:19:30.321423   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255170219684937	'<nil>'	'<nil>'	Apr 12 19:19:30.220119	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6714401Z W0412 19:19:30.559742   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255170219684943	'<nil>'	'<nil>'	Apr 12 19:19:30.459253	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6715682Z W0412 19:19:30.698749   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255170595050804	'<nil>'	'<nil>'	Apr 12 19:19:30.595315	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6716986Z W0412 19:19:30.801595   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255170699376957	'<nil>'	'<nil>'	Apr 12 19:19:30.699659	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6718265Z W0412 19:19:30.811860   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255170710196646	'<nil>'	'<nil>'	Apr 12 19:19:30.710569	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6737606Z W0412 19:19:30.914168   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255170812526882	'<nil>'	'<nil>'	Apr 12 19:19:30.812890	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6739555Z W0412 19:19:31.016365   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255170914918019	'<nil>'	'<nil>'	Apr 12 19:19:30.915175	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6741092Z W0412 19:19:31.027187   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255170925646707	'<nil>'	'<nil>'	Apr 12 19:19:30.926296	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6742445Z W0412 19:19:31.192838   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255171091410960	'<nil>'	'<nil>'	Apr 12 19:19:31.091750	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6743994Z W0412 19:19:31.295349   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255171193536093	'<nil>'	'<nil>'	Apr 12 19:19:31.193857	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6745492Z W0412 19:19:31.397516   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255171296040029	'<nil>'	'<nil>'	Apr 12 19:19:31.296310	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6746787Z W0412 19:19:31.561645   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255171460419269	'<nil>'	'<nil>'	Apr 12 19:19:31.460918	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6748361Z W0412 19:19:31.663729   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255171562175099	'<nil>'	'<nil>'	Apr 12 19:19:31.562370	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6749917Z W0412 19:19:31.766755   26560 tx_pool.go:126] killing transaction (exceeded timeout: 100ms): 1618255171664455533	'<nil>'	'<nil>'	Apr 12 19:19:31.664694	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6752035Z E0412 19:19:32.130507   26560 tx_executor.go:129] markFailed: UpdateRedo failed for dtid bb: unknown error: query: 'update _vt.redo_state set state = 0 where dtid = 'bb'' is not supported on fakesqldb (errno 1105) (sqlstate HY000) during query: update _vt.redo_state set state = 0 where dtid = 'bb'
2021-04-12T19:19:34.6754790Z E0412 19:19:32.134165   26560 tx_executor.go:129] markFailed: UpdateRedo failed for dtid aa: unknown error: query: 'update _vt.redo_state set state = 0 where dtid = 'aa'' is not supported on fakesqldb (errno 1105) (sqlstate HY000) during query: update _vt.redo_state set state = 0 where dtid = 'aa'
2021-04-12T19:19:34.6758665Z E0412 19:19:32.413259   26560 tx_engine.go:439] Error reading unresolved prepares: 'fake': unknown error: query: 'select count(*) from _vt.redo_state where time_created < 1618255169912901927' is not supported on fakesqldb (errno 1105) (sqlstate HY000) during query: select count(*) from _vt.redo_state where time_created < 1618255169912901927
2021-04-12T19:19:34.6760837Z W0412 19:19:33.423440   26560 stateful_connection_pool.go:96] killing transaction for shutdown: 1618255172423103411	'<nil>'	'<nil>'	Apr 12 19:19:32.423234	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6762287Z W0412 19:19:33.425445   26560 stateful_connection_pool.go:96] killing transaction for shutdown: 1618255173425220654	'<nil>'	'<nil>'	Apr 12 19:19:33.425429	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6763683Z W0412 19:19:33.428609   26560 tx_pool.go:126] killing transaction (exceeded timeout: 1ms): 1618255173427269771	'<nil>'	'<nil>'	Apr 12 19:19:33.427482	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6765054Z W0412 19:19:33.448363   26560 stateful_connection_pool.go:96] killing transaction for shutdown: 1618255173448018139	'<nil>'	'<nil>'	Apr 12 19:19:33.448341	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6766754Z W0412 19:19:34.450120   26560 tx_pool.go:126] killing transaction (exceeded timeout: 1s): 1618255173448818846	'principal:"principle" '	'username:"user" '	Apr 12 19:19:33.449627	Jan  1 00:00:00.000000	-9223372036.854776			
2021-04-12T19:19:34.6767517Z FAIL
2021-04-12T19:19:34.6768120Z FAIL	vitess.io/vitess/go/vt/vttablet/tabletserver	7.948s
```

### PR DESCRIPTION
**Pull request by vmg**
## Description

This is a flaky test that @harshit-gangal detected. I'm quite sure this is the proper fix.

## Related Issue(s)
Fixes #7832
Fixes #7837 

## Checklist
- [ ] Should this PR be backported?
- [ ] Tests were added or are not required
- [ ] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [x]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

**Review by harshit-gangal — COMMENTED** [review]
I tried this locally and it is not failing for the first part of the test.

It is failing below now.

```
--- FAIL: TestConsolidatorMemoryLimits (0.54s)
    --- FAIL: TestConsolidatorMemoryLimits/multiple_phase_consolidation (0.17s)
        stream_consolidator_test.go:398: 
            	Error Trace:	stream_consolidator_test.go:398
            	Error:      	Not equal: 
            	            	expected: 4
            	            	actual  : 3
            	Test:       	TestConsolidatorMemoryLimits/multiple_phase_consolidation

```

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vttablet/tabletserver/stream_consolidator_test.go [hunk 1]
```diff
@@ -97,6 +99,15 @@ func (ct *consolidationTest) leader(stream StreamCallback) error {
 	return nil
 }
 
+func (ct *consolidationTest) waitForResults(worker int, count int64) {
+	for {
+		if atomic.LoadInt64(&ct.results[worker].count) > count {
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/stream_consolidator_test.go:104]: Just making sure you intended for `>` and not `>=`?

Comment (vmg) [reply go/vt/vttablet/tabletserver/stream_consolidator_test.go:104]: Ayup, it needs to wait until we have received more than the expected amount of results.

**Review by shlomi-noach — APPROVED** [review]

**Comment by harshit-gangal** [comment]
@vmg rebase from master will fix CI failure.

**Pull request merged**